Add Docker support

buchdag opened the pull request:
This PR add Docker support to the project and the relevant Travis tests.

The image is based on Alpine linux 3.6 with Python 3 and weight ~ 85MB

The following:

```
services:
  - docker
```

has been removed from the travis.yml file as [it isn't needed anymore with the Travis Trusty build environment](https://docs.travis-ci.com/user/reference/trusty/#Docker) and triggered a warning message.

zenhack commented:
Sorry for the delay reviewing this. I'm going to block of some time Wednesday to pick through it.

buchdag commented:
No hurry 👍

zenhack commented on docker/entrypoint.sh:
```diff
@@ -0,0 +1,21 @@
+fi
+
+# else default to run whatever the user wanted like "bash"
+exec "$@"
```
This could be a bit simpler:

```bash
case "$1" in
	-*) exec simp_le $@ ;;
	'') exec simp_le ;;
	*) exec "$@"  ;;
esac
```

buchdag replied:
```
case "$1" in
	-*) exec simp_le "$@" ;;
	simp_le) shift; exec simp_le "$@" ;;
	*) exec "$@"  ;;
esac
```

Works as intended. The idea is to be able to run the container pretty much like you run the command (after the docker specific part), but remain able to bypass the default cmd of the container (`/usr/bin/simp_le --help` in this case):

`docker run zenhack/simp_le` will run simp_le with the --help argument.

`docker run zenhack/simp_le -v` and `docker run zenhack/simp_le simp_le -v` will both run simp_le with the -v argument.

`docker run -it zenhack/simp_le bash` will run an interactive bash session inside the container.

This is a prerequisite if you want to be able to base other containers on this one.

zenhack replied:
Yep, the interface is fine, I was mostly quibbling about the implementation, which I thought could be simpler.

In your version of the case statement, the second line seems superfluous. If it weren't there you'd just fall through to the one below it, which seems like it should do the same thing (rather than shifting away 'simp_le' and then adding it again, you just leave it there in the first place). Am I missing something?

buchdag replied:
You're not, the second line is indeed superfluous, I'll remove it.

buchdag commented:
Yes, the example does exactly the same thing as the command already provided at the top of the readme, but with the executable packaged in a docker container. Nothing fancy.

The container is basically an alternative way to install, use and/or test simp_le, as would be a Debian packaging, with advantages specific to Docker: 

A Docker container does not require any dependency to be installed on the host beside Docker itself (which you probably already have if you are reading how to use a given project through Docker) and the resulting container is a fully self contained unit. You don't want simp_le anymore ? Your just delete the container, nothing is left over.

I personally find it way easier and cleaner to do it that way than the bootstrap + virtualenv method, but I'm biased toward Docker as I use it heavily.

For testing purpose that makes it incredibly easy to have different versions coexisting on the same system (just change something then build the container again with a different tag). I've been testing everything I contributed so far this way.

Having a base container with a consistent interface that just package the executable allows to build upon it with another Dockerfile, for instance adding an embedded webserver like darkhttp to the image and tying it automatically to the default-webroot when the container is run (kinda like certbot ` --standalone` mode).

Does that sound like a useful addition to you ? If yes I'll try to expand the readme section.

zenhack commented:
Yeah, that sounds fine. I would just add a brief mention to the readme about easing installation.

buchdag commented:
Don't merge yet.

If you are ok with the PR as it is, I think It would be interesting to setup a Dockerhub repo with automated build from the GitHub repo. That would allow to pull pre-built images of zenhack/simp_le from Dockerhub instead of having to build them locally.

The Dockerfile would be moved to the /docker directory along with its own README.md file to be used as a description on Dockerhub, and the docker usage would be removed from README.rst

zenhack commented:
Works for me.

buchdag commented:
If you want to have a look at what it might look on Dockerhub: https://hub.docker.com/r/duch/simp_le/

Tagged version were built locally and then pushed to Dockerhub, but next versions can be built automatically on push the same way the `latest` version would be from `master`. It needs to be setup on Dockercloud rather than Dockerhub but the images end up on the same registry.

zenhack commented on README.rst:
```diff
@@ -76,6 +76,11 @@ Installation
+Usage with Docker
+-----------------
+
+If you want to use simp_le with Docker, have a look at https://github.com/zenhack/simp\_le/docker
```
This is a dead link (and still will be once it's merged); paths within the repo are like `/tree/master/docker`.

buchdag replied:
Will fix.

zenhack commented on docker/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+WORKDIR /simp_le/certs
+
+ENTRYPOINT [ "/bin/bash", "/simp_le/entrypoint.sh" ]
+CMD [ "--help" ]
```
I'm not super comfortable with this much copypasta; is there a way to factor some of this out?

buchdag replied:
Can you be more specific ? Is it the presence of two Dockerfiles or something inside this one ?

zenhack replied:
The fact that there are two (very similar) Dockerfiles. Sorry if that wasn't clear.

buchdag replied:
I have an idea how to combine the two, I'll test it and come back to you.

If it doesn't work we can stil have only one Dockerfile provided we use Dockercloud's automated build instead of Dockerhub's because of a missing feature in the latter.

The two alternatives have their own small downsides I'll detail later.

zenhack commented on docker/README.md:
````diff
@@ -0,0 +1,68 @@
+For testing and development purpose, you can build this image from the cloned GitHub repository:
+
+```
+$ git clone http://github.com/zenhack/simp_le.git
````
Should specify https here.

buchdag replied:
Will fix.

buchdag commented:
Ok, let me explain in details:

We'd like the Docker stuff to be placed in its own subdirectory (docker/) to avoid cluttering the rest of the project with files irrelevant to those that don't use Docker and to have a separate README.md for Dockerhub registry.

This in turn have an implication on the way docker build works. The build process context root is, by default, at the level of the Dockerfile being built. That means that you can't reach anything that is on a parent path to the Dockerfile. This can be overriden on the CLI but Dockerhub does not allow you to customize its build process this way.

Consequently, if we still want the Docker stuff to be on a subdirectory **and** have automated builds on Dockerhub, we have to bring the source into the build using git (or a derivative like pip install git+https://).

The resulting Dockerfile can't build from local sources, hence the second Dockerfile specifically made for this. Having multiple, slightly different Dockerfiles for different builds is a very common occurrence, actually official Docker-sponsored images are in fact [GitHub repositories full of Dockerfiles (organized by image variants and program versions)](https://github.com/nginxinc/docker-nginx/tree/master/mainline). I have a third non commited simp_le Dockerfile to build the image with python2 instead of python3.

Another option would be to use Dockercloud for the automated build. It has the required feature to change docker build context, so it can build from the local sources using the Dockerfile.localbuild, it still push to the image to the same place (Dockerhub registry) but you lose some of the Dockerhub automated build advantages, namely [displaying the Dockerfile](https://hub.docker.com/r/duch/letsencrypt-nginx-proxy-companion/~/dockerfile/), the [build history and details](https://hub.docker.com/r/duch/letsencrypt-nginx-proxy-companion/builds/), and the linked repository.

And if we only provide a Dockerfile building from source in a sub directory, it can't be used simply like `cd docker; docker build .`, you have to provide arguments for changing the build context. I'm not fond of providing a Dockerfile that can't be used in the most straightforward way.

As you see there are a lot of alternatives, none of them being "the best":

1. setup automated build on Dockerhub using the Dockerfile that grab sources with git, don't provide another Dockerfile, remove the ability to build from local sources.
2. setup automated build on Dockerhub using the Dockerfile that grab sources with git, provide a second Dockerfile to allow building from local sources (this is the current state)
3. setup automated build on Dockercloud using the Dockerfile for local sources, lose most of the Dockerhub automated build features, Dockerfile require additional arguments to be used.

I'm still in favor of 2) because as I explained multiple Docker files for different purposes is common and being able to build from local sources was the reason why I created this PR in the first place.

I binned the idea of a multipurpose Docker file as I realized just after writing it that it's a really "non Docker" idea. You don't want an all-in-one-do-it-all Dockerfile.

zenhack commented on docker/Dockerfile:
```diff
@@ -0,0 +1,21 @@
+
+RUN apk --no-cache add bash openssl python3 && \
+    apk --no-cache --virtual .build-deps add gcc git libffi-dev musl-dev openssl-dev py-pip python3-dev && \
+    pip3 install --no-cache-dir --upgrade pip setuptools wheel && \
```
Is this line actually needed on alpine? We reinstall/upgrade pip/setuptools/wheel in the venv.sh script because it's supposed to work on a range of distributions, including really stuff like rhel/centos 6.x which has outdated versions of these tools, but it isn't needed on most systems -- part of my gripe about the two Dockerfiles is that there's a certain amount of common boilerplate, getting rid of this would reduce that somewhat.

Also, could we encapsulate the bulk of this RUN command in a shell script, and call it from both Dockerfiles?

I've decided I'm okay having more than one, but I'd still like to factor out the common bits where we can.

buchdag replied:
I don't think the line is needed, on alpine 3.6 the only package pip actually update is wheel, from 28.8 to 29, I'll remove it.

I'm not convinced by the need to factor the common RUN commands between the two.

We'd be replacing this:

```
RUN apk --no-cache add bash openssl python3 && \
    apk --no-cache --virtual .build-deps add gcc git libffi-dev musl-dev openssl-dev py-pip python3-dev && \
    pip3 install --no-cache-dir git+https://github.com/zenhack/simp_le.git@$SIMP_LE_VERSION && \
    apk --update --purge del .build-deps
```

by thoses:

```
COPY .dependencies.sh /simp_le/dependencies.sh

RUN /simp_le/dependencies.sh install && \
    pip3 install --no-cache-dir git+https://github.com/zenhack/simp_le.git@$SIMP_LE_VERSION && \
    /simp_le/dependencies.sh purge && \
    rm /simp_le/dependencies.sh
```
```
#!/bin/bash

case "$1" in
    install)
    apk --no-cache add bash openssl python3 && \
    apk --no-cache --virtual .build-deps add gcc git libffi-dev musl-dev openssl-dev py-pip python3-dev
    ;;
    purge)
    apk --update --purge del .build-deps
    ;;
esac
```

Adding 12 more lines to factor out six (3 line times two Dockerfiles) does not look like a good deal to me. And as a side effect the resulting Dockerfile are really obfuscated.

zenhack replied:
I'd been thinking something more like:

```sh
set -e
apk --no-cache add bash openssl python3
apk --no-cache --virtual .build-deps add gcc git libffi-dev musl-dev openssl-dev py-pip python3-dev
pip3 install --no-cache-dir "$1"
[ -d "$1" ] && rm -rf "$1"
apk --update --purge del .build-deps
```
```
COPY .docker-deps.sh /simp_le/deps.sh
RUN /simp_le/deps.sh git+https://github.com/zenhack/simp_le.git@$SIMP_LE_VERSION
```

```
COPY .docker-deps.sh /simp_le/deps.sh
RUN /simp_le/deps.sh /simp_le/src
```

Thoughts?

buchdag replied:
I've got both Dockerfiles working with that solution but I really have issues with the resulting obfucated Dockerfile that will appear on Dockerhub. You won't have any idea of what it is doing unless you go check the install script on Github.

[Judging](https://github.com/docker-library/python/blob/master/3.6/alpine3.6/Dockerfile) [from](https://github.com/docker-library/mariadb/blob/master/10.3/Dockerfile) [official](https://github.com/docker-library/wordpress/blob/master/php7.1/apache/Dockerfile) [images](https://github.com/docker-library/redis/blob/ebde981d2737c5a618481f766d253afff13aeb9f/3.2/Dockerfile),  from what I know of Docker and how I used/learned it so far, I really don't think hiding Dockerfile's RUN commands behind wrapper script is considered a good practice. On the contrary it seems common practice to have them be extremely clear and explicit about what they do, at the expanse of redundancy. I completely understand the drive to get rid of redundant code, but I think in this case it's missing the point.

If you really aren't ok with not factoring out the common code between the two, I'd prefer to remove the local build Dockerfile and find another way to let people do this (detailing it in the doc maybe ?).

Sorry if you feel I'm being annoying on this one.

zenhack replied:
No, it's fine, I think it's worth hashing stuff like this out -- I am not annoyed in the least.

I'm willing to buy the argument "this is what idiomatic docker looks like."

My own feeling is the shell script I presented isn't terribly obscure/complicated, and the dockerhub links to the github repo, so it would be easy enough to just look at that file. But I'm also coming from a professional software engineering background, maybe I'm under-estimating how hard it would be to follow for someone who has more of an ops background, which I imagine is a decent chunk of docker's user base these days.

In any case, I'm okay leaving this duplicated. Make the other changes we talked about and I'll merge.

Thanks again for the patch!

buchdag commented:
Every requested changes should be in there:

- corrected the relative link in README.rst
- changed an `http` to `https` on README.md
- removed the pip upgrade of pip, setuptools and wheel

In addition:

- corrected the image size in README.md, it was off by 10 MB
- made the link to the Manifesto a bit less dry than just the URL
- prettyfied a bit the `apk add` commands
- added one last rm command to squeeze another 1.1 MB out of the image

zenhack commented on README.rst:
```diff
@@ -88,6 +93,7 @@ Freenode)`_
 .. _Let’s Encrypt: https://letsencrypt.org
 .. _UNIX philosophy: https://en.wikipedia.org/wiki/Unix_philosophy
 .. _IRC (#simp\_le at Freenode): http://webchat.freenode.net?randomnick=1&channels=%23simp_le&prompt=1
+.. _simp\_le for Docker: docker
```
This doesn't render correctly; it tries to link to a blob (regular file) called docker, but it's a directory. You probably want the README.

buchdag replied:
I intended to link to the directory (to see both the readme and the files when you clink on the link), it appeared to work when I tried it, is this actually something to avoid ?

On a sidenote it seems escaping the underscores isn't needed on GitHub.

zenhack commented:
Quoting Nicolas Duchon (2017-09-18 03:16:22)

>    I intended to link to the directory (to see both the readme and the
>    files when you clink on the link), it appeared to work when I tried it,
>    is this actually something to avoid ?


Okay, yeah it seems to work when browsing the repository; must be
something odd to do with the way it's rendered from the diff. Merged,
thanks again for the patch.

>    On a sidenote it seems escaping the underscores isn't needed on GitHub.

Yeah, I didn't write that, so can't speak to why some of them are like
that; maybe it causes problems in other contexts where RST is used? I
haven't bothered to look into it.

buchdag commented:
Let me know if you need help with Dockerhub.

> maybe it causes problems in other contexts where RST is used?

[Yep, it appears to comes from differences in markdown implementations.](https://stackoverflow.com/questions/10302516/markdown-can-i-have-underscores-w-o-escaping-them-and-not-have-markdown-italic)
